contracts/Gateway/NativeTokenGateway.sol

```diff
@@ -0,0 +1,204 @@
+     * @param minter The address on behalf of whom the supply is performed.
+     * @custom:error ZeroAddressNotAllowed is thrown if either vToken or minter address is zero address
+     * @custom:error ZeroValueNotAllowed is thrown if mintAmount is zero
+     * @custom:event TokensWrappedAndSupplied is emitted when assets are supplied to the market
+     */
+    function wrapAndSupply(IVToken vToken, address minter) external payable nonReentrant {
+        ensureNonzeroAddress(address(vToken));
+        ensureNonzeroAddress(minter);
+
+        uint256 mintAmount = msg.value;
+        ensureNonzeroValue(mintAmount);
+
+        wrappedNativeToken.deposit{ value: mintAmount }();
+        wrappedNativeToken.approve(address(vToken), mintAmount);
+
+        vToken.mintBehalf(minter, mintAmount);
+
+        wrappedNativeToken.approve(address(vToken), 0);
+        emit TokensWrappedAndSupplied(minter, address(vToken), mintAmount);
+    }
+
+    /**
+     * @notice Redeem vWETH, unwrap to ETH, and send to the user
+     * @param vToken The vToken market to interact with
+     * @param redeemAmount The amount of underlying tokens to redeem
+     * @custom:event TokensRedeemedAndUnwrapped is emitted when assets are redeemed from a market and unwrapped
+     */
+    function redeemUnderlyingAndUnwrap(IVToken vToken, uint256 redeemAmount) external nonReentrant {
+        ensureNonzeroAddress(address(vToken));
+        ensureNonzeroValue(redeemAmount);
+
+        uint256 balanceBefore = wrappedNativeToken.balanceOf(address(this));
+        vToken.redeemUnderlyingBehalf(msg.sender, redeemAmount);
+        uint256 balanceAfter = wrappedNativeToken.balanceOf(address(this));
+
+        uint256 nativeTokenBalanceBefore = address(this).balance;
+        wrappedNativeToken.withdraw(balanceAfter - balanceBefore);
+        uint256 nativeTokenBalanceAfter = address(this).balance;
+
+        uint256 redeemedAmount = nativeTokenBalanceAfter - nativeTokenBalanceBefore;
+
+        _safeTransferETH(msg.sender, redeemedAmount);
+        emit TokensRedeemedAndUnwrapped(msg.sender, address(vToken), redeemedAmount);
+    }
+
+    /**
+     * @dev Borrow wrappedNativeToken, unwrap to Native, and send to the user
+     * @param vToken The vToken market to interact with
+     * @param borrowAmount The amount of underlying tokens to borrow
+     * @custom:error ZeroAddressNotAllowed is thrown if  vToken address is zero address
+     * @custom:error ZeroValueNotAllowed is thrown if borrowAmount is zero
+     * @custom:event TokensBorrowedAndUnwrapped is emitted when assets are borrowed from a market and unwrapped
+     */
+    function borrowAndUnwrap(IVToken vToken, uint256 borrowAmount) external nonReentrant {
+        ensureNonzeroAddress(address(vToken));
+        ensureNonzeroValue(borrowAmount);
+
+        vToken.borrowBehalf(msg.sender, borrowAmount);
+
+        wrappedNativeToken.withdraw(borrowAmount);
+        _safeTransferETH(msg.sender, borrowAmount);
```

Comment:
shouldn't we check for the balanceBefore and balaceAfter as we doing in other functions

Reply:
There's no need to compare the balanceBefore and balanceAfter, as the contract's **WNative** balance will increase precisely by the user-specified borrowAmount. This is because, upon a successful **borrowBehalf** call, we transfer the exact input amount to the user. Similarly, when invoking the **WNative** contract's withdraw function, it disburses the exact amount specified, which will be the borrowAmount.